### ISSUE TITLE
Interpreting the Cobertura results

### ISSUE THREAD

**Issue by jonassvalin** [tntim96/JSCover#199]
Hello, and thanks in advance for this very helpful tool. We are currently working with JSCover on a large scale development project and we've successfully integrated the Cobertura results into our Jenkins build process. We are having some issues interpreting the results however, here is an example of how they can look for a test case:
![coverageexample](https://cloud.githubusercontent.com/assets/8707399/8565816/51435f2e-2559-11e5-814d-b3bc5db3e0ee.png)

The report indicates that 2 out 2 packages have been covered, and 4 out of 4 files have been covered. While it's true that 4 files have been covered, there are far more than 4 total files in the project (number of files are in the hundreds, and packages in the dozens), so it should be 4 out of XXX, right? Similary lines covered says 6 out of 945, but 945 is only the total for the covered files, not the entire project.

Is there something you can immedietly point to that we're doing wrong which causes this, or are we interpreting the results incorrectly? We seem to be referencing the source code correctly when creating the Cobertura XML, because if we go into the "breakdown by file" it correctly displays the source code of the covered files atleast. 
Thanks in advance

**Comment by jonassvalin** [comment]
https://github.com/tntim96/JSCover/issues/39

Just checked and realized we had not updated our JSCover version (were using com.github.tntim96:JSCover:1.0.17) recently and this might have been fixed already per the above post. I will re-run the build and test suite and get back to you.

**Comment by tntim96** [comment]
Nothing in that area has changed since 1.0.17, so it could be a problem still.

> While it's true that 4 files have been covered, there are far more than 4 total files in the project

Are you sure all the files are either loaded, or that you are using `--include-unloaded-js`?


If the above doesn't fix the issue, maybe you could post the `jscoverage.json` file somewhere and I can test what's going on with the Cobertura conversion.

**Comment by jonassvalin** [comment]
We are using File Mode, so we instrument all the code beforehand and the results are exported converted to cobertura at the end of the test. If i understand the manual correctly the --include-unloaded-js option is only available for Server Mode?

**Comment by tntim96** [comment]
That's correct, so for JSCover to know about the missing files, they'll have to be loaded by the tests.

If that's not possible, I guess JSCover could be enhanced to generate an empty `jscoverage.json` file you can merge subsequent results to so that all the files are included.

**Comment by jonassvalin** [comment]
All right, i understand. It is not possible for us to adapt the tests to do this currently. Ideally we would like an option when creating the Cobertura XML that includes unloaded files. This could theoretically be done by creating a jscoverage.json that includes all files (but none having any lines executed) and merging this with the actual coverage results before creating the Cobertura summary? Such a solution would serve our needs perfectly, but perhaps it is inconvenient to implement?



(On an unrelated note, me and a colleague used JSCover as part of an automated fault localization tool that we developed for our master's thesis, if you're interested in looking at the report feel free to email me at: jonassvalin(at)gmail(dot)com and i'll send it over)

**Comment by tntim96** [comment]
It shouldn't be too hard to add and will start working on it. I should hopefully have something ready to test soon.

**Comment by jonassvalin** [comment]
That's great, thank you. We will test it as soon as it's available.

**Comment by tntim96** [comment]
Before I commit, can you test with either [this test JAR](http://jscover.sourceforge.net/JSCover-all.jar) or the [1.0.20-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/com/github/tntim96/JSCover/1.0.20-SNAPSHOT/) which will need this in your POM
```xml
<pluginRepository>
    <id>Sonatype repository</id>
    <name>Sonatype's Maven repository</name>
    <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
</pluginRepository>
```

The switch to use is `--include-unloaded-js`.

**Comment by jonassvalin** [comment]
Is it possible to make it available through gradle? It's what we use in our build process. I tried instrumenting locally and it worked just fine, but unfortunately I can't build the product and run the tests locally.

**Comment by tntim96** [comment]
Gradle uses a similar dependency management to Maven. I think it would be something like this:
```
dependencies {
    compile("com.github.tntim96:JSCover:1.0.20-SNAPSHOT")
}
repositories {
    mavenCentral()
    maven {
        url 'https://oss.sonatype.org/content/repositories/snapshots'
    }
}
```

**Comment by jonassvalin** [comment]
Excellent, that worked fine locally so I will try it in Jenkins and verify the test results as soon as they're available.

**Comment by jonassvalin** [comment]
Having some complications, unsure what is wrong. Retrieving JSCover:1.0.20-SNAPSHOT using gradle works fine and instrumenting normally also works fine, but when the --include-unloaded-js command is added JSCover complains that the command is not recognized. 

It works fine locally with the JSCover-all.jar that you linked however.

**Comment by tntim96** [comment]
I've just re-uploaded the SNAPSHOTS and checked them, they should work. Can you retry?

**Comment by jonassvalin** [comment]
I retried and it still fails with the same output. If i try the exact same command locally with the downloaded jar it works fine. 

What is strang is if i run -V on the downloaded jar and in the gradle call they both claim that the version is the correct one, 1.0.20-SNAPSHOT.

**Comment by tntim96** [comment]
Can you try forcing gradle to get the latest SNAPSHOT as described [here](https://discuss.gradle.org/t/how-to-get-gradle-to-download-newer-snapshots-to-gradle-cache-when-using-an-ivy-repository/7344).

Also, you can try deleting the SNAPSHOTs from your local repository.

**Comment by jonassvalin** [comment]
Yes, that worked! I will get back to you once we've tested it in Jenkins.

**Comment by tntim96** [comment]
Great - thanks for re-trying

**Comment by jonassvalin** [comment]
We got it running in Jenkins, but unfortunately it does not seem to work, we got the exact same coverage results as before. Does it matter that we're using the "return jscoverage_serializeCoverageToJSON()" command rather than "jscoverage_report()"?

**Comment by tntim96** [comment]
So when you instrument the JS in file-mode with the `--include-unloaded-js`, an empty `jscoverage.json` file is created. If you inspect the JSON, you should be able to tell if all the relevant files are included.

If they're there, you need to merge this file with the other one you are generating.

**Comment by jonassvalin** [comment]
Right, i thought it would be automatically added when running the command. Merging them manually is actually much better because it gives us more freedom. I will verify it first thing monday.

**Comment by tntim96** [comment]
Great.

I've released v1.0.20 anyway. I can do another release if there are any issues.

**Comment by jonassvalin** [comment]
Fix seems to work just fine, I will re-open if I find any issues. Many thanks for quickly adding this functionality!
